control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -128,6 +152,47 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return ctrl.Result{}, err
 	}
 
+	// Return early if deleted
+	if !hostedControlPlane.DeletionTimestamp.IsZero() {
+		if err := r.delete(ctx, hostedControlPlane); err != nil {
+			r.Log.Error(err, "failed to delete cluster")
+			return ctrl.Result{}, err
+		}
+		if controllerutil.ContainsFinalizer(hostedControlPlane, finalizer) {
+			controllerutil.RemoveFinalizer(hostedControlPlane, finalizer)
+			if err := r.Update(ctx, hostedControlPlane); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from cluster: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	patchHelper, err := patch.NewHelper(hostedControlPlane, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to init patch helper: %w", err)
+	}
+
+	oldStatus := hostedControlPlane.Status.DeepCopy()
+
+	// Sync any status changes during the reconciliation
+	defer func() {
+		availableCondition := getAvailableCondition(hostedControlPlane)
+
+		if len(oldStatus.Conditions) != 0 {
+			oldAvailableCondition := &oldStatus.Conditions[0]
+			if reflect.DeepEqual(oldAvailableCondition, availableCondition) {
+				// No change to status
+				return
+			}
+		}
+		availableCondition.LastTransitionTime = metav1.Now()
+
+		if err := patchHelper.Patch(ctx, hostedControlPlane); err != nil {
+			r.Log.Error(err, "failed to patch")
```

Comment:
Should this requeue?

Reply:
unfortunately, we can't requeue from the defer.  if we could, we probably should.

Reply:
So may be we want to call this method explicitly rather than deferring to avoid potential side effects of failing to patch. Although the code will be more verbose we'll gain on predictability and robusteness.

Reply:
Agree, I think it makes sense to not do it in a defer